### PR TITLE
feat: udsink bidirectional streaming

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: src/main/java/io/numaproj/numaflow/sinker/Service.java [hunk 1]
```diff
@@ -41,25 +34,58 @@ public Service(Sinker sinker) {
...
-        DatumIteratorImpl datumStream = new DatumIteratorImpl();
+            @Override
+            public void onNext(SinkOuterClass.SinkRequest request) {
+                // make sure the handshake is done before processing the messages
+                if (!handshakeDone) {
+                    if (!request.getHandshake().getSot()) {
+                        responseObserver.onError(new Exception("Handshake request not received"));
+                        return;
+                    }
+                    responseObserver.onNext(SinkOuterClass.SinkResponse.newBuilder()
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/sinker/Service.java:51]: In go sdk, we also set result status to success.

File: src/main/java/io/numaproj/numaflow/sinker/Service.java [hunk 2]
```diff
@@ -41,25 +34,58 @@ public Service(Sinker sinker) {
...
+                }
 
-        Future<ResponseList> result = sinkTaskExecutor.submit(() -> this.sinker.processMessages(
-                datumStream));
+                // Create a DatumIterator to write the messages to the sinker
+                // and start the sinker if it is the start of the stream
+                if (startOfStream) {
+                    datumStream = new DatumIteratorImpl();
+                    result = CompletableFuture.supplyAsync(
+                            () -> sinker.processMessages(datumStream),
```
Comment (RohanAshar) [inline src/main/java/io/numaproj/numaflow/sinker/Service.java:63]: How are we planning to handle error handling with Future failing?

Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/sinker/Service.java:63]: there is a completion exception catch block, which returns error back to the client

File: src/main/java/io/numaproj/numaflow/sourcer/Service.java [hunk 3]
```diff
@@ -89,16 +93,19 @@ public void onCompleted() {
...
     public StreamObserver<SourceOuterClass.AckRequest> ackFn(final StreamObserver<SourceOuterClass.AckResponse> responseObserver) {
         return new StreamObserver<>() {
+            private boolean handshakeDone = false;
             @Override
             public void onNext(SourceOuterClass.AckRequest request) {
-                // if the request is a handshake, send a handshake response
-                if (request.hasHandshake() && request.getHandshake().getSot()) {
+                // make sure that the handshake is done before processing the ack requests
+                if (!handshakeDone) {
+                    if (!request.getHandshake().getSot()) {
```
[the remark this answers is not in the document]
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/sourcer/Service.java]: added it back.

**Review by KeranYang — APPROVED** [review]

File: src/main/java/io/numaproj/numaflow/sinker/SinkerTestKit.java [hunk 4]
```diff
@@ -113,11 +113,13 @@ public Client(String host, int port) {
          * @return response from the server as a ResponseList
          */
         public ResponseList sendRequest(DatumIterator datumIterator) {
-            CompletableFuture<SinkOuterClass.SinkResponse> future = new CompletableFuture<>();
```
Comment (Krithika3) [inline src/main/java/io/numaproj/numaflow/sinker/SinkerTestKit.java:116]: just curious, as to why SinkerTestKit lives in this package?

File: src/main/java/io/numaproj/numaflow/sinker/SinkerTestKit.java [hunk 5]
```diff
@@ -168,28 +174,39 @@ public void onCompleted() {
...
                 throw new RuntimeException(e);
             }
 
             ResponseList.ResponseListBuilder responseListBuilder = ResponseList.newBuilder();
-            for (SinkOuterClass.SinkResponse.Result result : response.getResultsList()) {
-                if (result.getStatus() == SinkOuterClass.Status.SUCCESS) {
-                    responseListBuilder.addResponse(Response.responseOK(result.getId()));
-                } else if (result.getStatus() == SinkOuterClass.Status.FALLBACK) {
+            for (SinkOuterClass.SinkResponse result : outputResponses) {
+                if (result.getHandshake().getSot()) {
```
Comment (Krithika3) [inline src/main/java/io/numaproj/numaflow/sinker/SinkerTestKit.java:197]: what does getSot do?

Comment (KeranYang) [reply src/main/java/io/numaproj/numaflow/sinker/SinkerTestKit.java:197]: sot means start of the transmission. when set to true, it means the handshake is successful.

**Pull request merged**